### PR TITLE
feat(protocol): Compressors

### PR DESCRIPTION
**Pull request by refcell**
### Description

Introduces compressors into `op-alloy-protocol`, cleaning up the brotli and zlib utility methods as well as the `ChannelOut` type.

### CODE REVIEW

**Comment by refcell** [comment]
## 📚 $\text{Stack Overview}$

Pulls submitted in this stack:
* #299 👈
* #298
* #296
* `main`

_This comment was automatically generated by [`st`](https://github.com/clabby/st)._

File: crates/protocol/src/channel_out.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    /// The encoded batch exceeds the max RLP bytes per channel.
+    #[error("The encoded batch exceeds the max RLP bytes per channel")]
+    ExceedsMaxRlpBytesPerChannel,
+}
+
+/// [ChannelOut] constructs a channel from compressed, encoded batch data.
+#[derive(Debug, Clone)]
+pub struct ChannelOut<'a, C>
+where
+    C: Compressor + Clone + core::fmt::Debug,
```
Comment (mattsse) [inline crates/protocol/src/channel_out.rs:38]: I think you can relax this here entirely

File: crates/protocol/src/channel_out.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    pub frame_number: u16,
+    /// Compressed batch data.
+    pub compressed: Option<Bytes>,
+    /// The compressor.
+    pub compressor: C,
+}
+
+impl<'a, C> ChannelOut<'a, C>
+where
+    C: Compressor + Clone + core::fmt::Debug,
```
Comment (mattsse) [inline crates/protocol/src/channel_out.rs:60]: is clone + debug required here?

File: crates/protocol/src/compression/zlib.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+use crate::Compressor;
+use alloc::vec::Vec;
+use miniz_oxide::inflate::DecompressError;
+
+/// The best compression.
+const BEST_ZLIB_COMPRESSION: u8 = 9;
+
+/// The ZLIB compressor.
+#[derive(Debug, Clone)]
+pub struct ZlibCompressor;
```
Comment (mattsse) [inline crates/protocol/src/compression/zlib.rs:12]:
```suggestion
#[derive(Debug, Clone, Default)]
#[non_exhaustive]
pub struct ZlibCompressor;
```

**Review by mattsse — APPROVED** [review]
lgtm

File: crates/protocol/src/channel_out.rs (hunk 2 above)
Comment (refcell) [reply crates/protocol/src/channel_out.rs:60]: I think so in order for the struct to be `Clone + Debug` since it's composition

**Pull request merged**